sdk/go/pulumi/context.go

```diff
@@ -2786,6 +2787,23 @@ func (ctx *Context) NewOutput() (Output, func(interface{}), func(error)) {
 	return newAnyOutput(&ctx.state.join)
 }
 
+// getSourcePositionForFrame converts a runtime.Frame to an RPC source position. Returns nil if the frame is not valid.
+func (ctx *Context) getSourcePositionForFrame(frame runtime.Frame) *pulumirpc.SourcePosition {
+	if frame.File == "" || frame.Line == 0 {
+		return nil
+	}
+
+	line := int32(-1)
+	if frame.Line <= math.MaxInt32 {
+		//nolint:gosec
+		line = int32(frame.Line)
+	}
+	return &pulumirpc.SourcePosition{
+		Uri:  "file:///" + path.Join(slices.Collect(fxs.Map(strings.Split(frame.File, "/"), url.PathEscape))...),
```

Comment:
Mind using the [helper](https://github.com/pulumi/pulumi/blob/33a8e42bad928815e90e8a82869fbecbdef38b7b/sdk/go/common/slice/slice.go#L28) we already have?

```suggestion
		Uri:  "file:///" + path.Join(slice.Map(strings.Split(frame.File, "/"), url.PathEscape)...),
```